File: 1171.php
<?php

/**
 * Definition for a singly-linked list.
 * class ListNode {
 *     public $val = 0;
 *     public $next = null;
 *     function __construct($val) { $this->val = $val; }
 * }
 */
class Solution {

    /**
     * @param ListNode $head
     * @return ListNode
     */
    function removeZeroSumSublists($head) {
        if (!$head) {
            return $head;
        }
        $dummy = new ListNode(null);
        $dummy->next = $head;
        $pre = $dummy;
        while ($pre) {
            $cur = $pre->next;
            $sum = 0;
            while ($cur) {
                $sum += $cur->val;
                $cur = $cur->next;

                if ($sum == 0) {
                    $pre->next = $cur;
                    break;
                }
            }
            if (!$cur) $pre = $pre->next;
        }
        return $dummy->next;
    }
}

File: go/linkedList/1171.go
package main

//func removeZeroSumSublists(head *ListNode) *ListNode {
//	if head == nil {
//		return head
//	}
//	dummy := new(ListNode)
//	dummy.Next = head
//
//	current := dummy
//	for current != nil {
//		pre, sum := current.Next, 0
//
//		for pre != nil {
//			sum += pre.Val
//			pre = pre.Next
//			if sum == 0 {
//				current.Next = pre
//				break
//			}
//		}
//		if pre == nil {
//			current = current.Next
//		}
//	}
//
//	return dummy.Next
//}
//
//func main() {
//
//}
//
//type ListNode struct {
//	Val  int
//	Next *ListNode
//}
